first commit for CT witness (#850)
google/certificate-transparency-go: witness/witness/README.md

CT Witness
==============

The witness is an HTTP service that stores STHs it has seen from
a configurable list of Certificate Transparency logs in a sqlite database.  This 
is a lightweight way to help detect or even prevent split-view attacks.  An 
overview of witnessing can be found in 
[trillian-examples](https://github.com/google/trillian-examples/tree/master/witness), along with "generic" witness implementations.  This witness is designed to be 
compatible with the specific formats used by CT.

Once up and running, the witness provides three API endpoints (as defined in
[api/http.go](api/http.go)):

> - `/witness/v0/logs` returns a list of all logs for which the witness is

Author: Sure, done.

  currently storing an STH.
- `/witness/v0/logs/<logid>/update` acts to update the STH stored for `logid`.
- `/witness/v0/logs/<logid>/sth` returns the latest STH for `logid`.

Running the witness
--------------------

Running the witness is as simple as running `go run main.go` (where `main.go`
can be found in the `cmd/witness` directory), with the following flags:
- `listen`, which specifies the address and port to listen on.
- `db_file`, which specifies the desired location of the sqlite database.  The
  use of sqlite limits the scalability and reliability of the witness (because
  this is a local file), so if that is required a different database backend
  would be needed.
- `config_file`, which specifies configuration information for the logs.  A 

>   sample configuration file is at `cmd/witness/example.conf`, and in general it

Author: Sure, added.

CT Witness
==============

The witness is an HTTP service that stores STHs it has seen from
a configurable list of Certificate Transparency logs in a sqlite database.  This 
is a lightweight way to help detect or even prevent split-view attacks.  An 
overview of witnessing can be found in 
[trillian-examples](https://github.com/google/trillian-examples/tree/master/witness), 
along with "generic" witness implementations.  This witness is designed to be 
compatible with the specific formats used by CT.

Once up and running, the witness provides three API endpoints (as defined in
[api/http.go](api/http.go)):
- `/ctwitness/v0/logs` returns a list of all logs for which the witness is
  currently storing an STH.
- `/ctwitness/v0/logs/<logid>/update` acts to update the STH stored for `logid`.
- `/ctwitness/v0/logs/<logid>/sth` returns the latest STH for `logid`.

Running the witness
--------------------

> Running the witness is as simple as running `go run main.go` (where `main.go`
> can be found in the `cmd/witness` directory), with the following flags:

Reviewer: ```suggestion
Running the witness is as simple as running `go run ./cmd/witness` from this directory, with the following flags:
```

Author: Accepted!

- `listen`, which specifies the address and port to listen on.
- `db_file`, which specifies the desired location of the sqlite database.  The
  use of sqlite limits the scalability and reliability of the witness (because
  this is a local file), so if that is required a different database backend
  would be needed.
- `config_file`, which specifies configuration information for the logs.  This
  repository contains a [sample configuration file](cmd/witness/example.conf), 
  and in general it is necessary to specify the following fields for each log:
    - `logID`, which is the identifier for the log.
    - `pubKey`, which is the public key of the log.

> - `private_key`, which specifies the private signing key of the witness.

Reviewer: This should be clear on the format, and whether it's taking the key raw, or a file path to it. Without looking through the code I don't know if this a filepath to a PEM / DER, or a note-style one-liner string.

Author: At the moment the format doesn't matter, since the witness never uses its secret key :).  I've added a line making that more clear.
